[DAG] Add generic expansion for ISD::FCANONICALIZE nodes

dominik-steenken opened the pull request:
This PR takes the work previously done by @pawan-nirpal-031 on X86 in #106370, and makes it available in common code. This should enable all targets to use `__builtin_canonicalize`  for all `f(16|32|64|128)` data types.

Canonicalization is implemented here as multiplication by `1.0`, as suggested in [the docs](https://llvm.org/docs/LangRef.html#llvm-canonicalize-intrinsic).

Moving the X86 implementation to common code was suggested in the reviews of #106370 by @arsenm .

@pawan-nirpal-031 I hope you don't mind that i went ahead  with this to get it working on SystemZ.

I don't have a lot of experience with this part of the code base yet, so reviews are very welcome.

dominik-steenken committed:
[SystemZ] Add tests for llvm.canonicalize.f(16|32|64|128).

This new test file essentially checks the code generation for the common code
expansion of `ISD::FCANONICALIZE`, since SystemZ has no custom handling
for it at this time. The tests are heavily based on the equivalent tests
for X86.

dominik-steenken committed:
[DAG] Implement generic llvm.canonicalize.f(16|32|64|128).

Previously, this expansion for the intrinsic has been implemented in the X86
backend. As suggested in the comments for the PR that merges that
implementation (#106370), this commit replicates that implementation (at least
for `f(16|32|64|128)`) in common code.

Canonicalization is thus implemented here as multiplication with 1.0.

This commit also touches the following backends: AArch64, AMDGPU, Mips,
PowerPC, and LoongArch. These backends previously added their own legalization
for `ISD::FCANONICALIZE`, but did not set the OperationAction accordingly,
since there was no default OperationAction in common code. Thus, this commit
adds the required `setOperationAction` calls and is otherwise intended as NFC
for these backends.

dominik-steenken commented:
@uweigand FYI

pawan-nirpal-031 commented:
Hi @dominik-steenken absolutely, Please go ahead with this, I have been wanting to get back and finish on the commitments as following PRs, this is much welcome. I will make some time to support canonicalization of denormals ( I have a PR just haven't been able to attend to it ) Thanks for your PR.

dominik-steenken commented:
Ok, so i updated the comment as requested, got rid of the `undef` check in order to remove the error from the `undef` deprecator and added a bunch of `setOperationAction(..., Legal)` to various backends that had independently handled `fcanonicalize` via their Tablegen files in order to get the tests clean again. Would appreciate another review, especially from folks of the backends that i had to touch in order to fix the tests :).

arsenm commented:
Patch title is now less clear, this is DAG codegen support

dominik-steenken commented:
> Patch title is now less clear, this is DAG codegen support

So, something like "DAG codegen support for `llvm.canonicalize`"?

RKSimon commented on llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp:
```diff
@@ -3356,6 +3356,27 @@ bool SelectionDAGLegalize::ExpandNode(SDNode *Node) {
   }
+  case ISD::FCANONICALIZE: {
+    // This implements llvm.canonicalize.f* by multiplication with 1.0,
+    // as suggested in https://llvm.org/docs/LangRef.html#id2335. It uses
```
can we use `https://llvm.org/docs/LangRef.html#llvm-canonicalize-intrinsic` instead? id### anchors will bitrot

dominik-steenken replied:
sure

dominik-steenken commented on llvm/test/CodeGen/SystemZ/canonicalize-vars.ll:
```diff
@@ -0,0 +1,131 @@
+  ret double %canonicalized
+}
+
+define void @canonicalize_ptr_f32(float * %out) {
```
ok, i've replaced `float *`, `double *`, `<4 x float> *` and <4 x double> *` with `ptr`.

dominik-steenken commented on llvm/test/CodeGen/SystemZ/canonicalize-vars.ll:
```diff
@@ -0,0 +1,131 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --default-march s390x-unknown-linux-gnu --version 5
+; RUN: llc -mtriple=s390x-linux-gnu -mcpu=z16 < %s | FileCheck %s -check-prefixes=Z16
+
+define float @canonicalize_fp32(float %a) {
```
I have added half and extended fp tests. From what i understand, the other touched targets had already implemented their own legalization of `fcanonicalize`, and would have added tests for all of their supported floating point types then. This PR should not change anything there - the targets only needed to be touched since they assumed that common code would not expand by default.

dominik-steenken commented on llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp:
```diff
@@ -3356,6 +3356,28 @@ bool SelectionDAGLegalize::ExpandNode(SDNode *Node) {
+    // Produce multiplication node x * 1.0.
+    SDValue Chain = DAG.getEntryNode();
+    SDValue Mul = DAG.getNode(ISD::STRICT_FMUL, dl, {VT, MVT::Other},
+                              {Chain, Operand, One});
```
I'm now propagating the flags and have added `NoFPExcept`.

dominik-steenken commented:
I believe the CI check fails were issues introduced into `main` by https://github.com/llvm/llvm-project/pull/146048 and https://github.com/llvm/llvm-project/pull/146247, but which should be fixed by now. I have rebased and the branch, CI is running.

RKSimon commented on llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp:
```diff
@@ -3356,6 +3356,28 @@ bool SelectionDAGLegalize::ExpandNode(SDNode *Node) {
+    SDValue Mul = DAG.getNode(ISD::STRICT_FMUL, dl, {VT, MVT::Other},
+                              {Chain, Operand, One});
+
+    Results.push_back(Mul);
```
@dominik-steenken Add a comment mentioning that ignoring the output chain was intended.

arsenm replied:
lgtm but should include this comment about the chain

RKSimon commented on llvm/test/CodeGen/SystemZ/canonicalize-vars.ll:
```diff
@@ -0,0 +1,437 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --default-march s390x-unknown-linux-gnu --version 5
+; RUN: llc -mtriple=s390x-linux-gnu -mcpu=z16 < %s | FileCheck %s -check-prefixes=Z16
+
+define half @canonicalize_fp16(half %a) {
```
use `) nounwind {` to hide all the cfi noise from all the tests

dominik-steenken replied:
i have added `nounwind`  to all the `fp16` tests.

RKSimon reviewed (commented):
A couple of minors but otherwise LGTM @arsenm ?

RKSimon commented on llvm/test/CodeGen/SystemZ/canonicalize-vars.ll:
```diff
@@ -0,0 +1,407 @@
+  ret void
+}
+
+ nounwind
```
typo (and cause of the CI failure)

RKSimon reviewed (approved):
LGTM - @arsenm any more comments?

RKSimon committed:
Merge branch 'main' into generic-canonicalize
